app/src/main/java/org/schabi/newpipe/util/potoken/PoTokenProviderImpl.kt

```diff
@@ -48,56 +49,52 @@ object PoTokenProviderImpl : PoTokenProvider {
      * case the current [webPoTokenGenerator] threw an error last time
      * [PoTokenGenerator.generatePoToken] was called
      */
-    private fun getWebClientPoToken(videoId: String, forceRecreate: Boolean): PoTokenResult {
+    private suspend fun getWebClientPoToken(videoId: String, forceRecreate: Boolean): PoTokenResult {
         // just a helper class since Kotlin does not have builtin support for 4-tuples
         data class Quadruple<T1, T2, T3, T4>(val t1: T1, val t2: T2, val t3: T3, val t4: T4)
 
         val (poTokenGenerator, visitorData, streamingPot, hasBeenRecreated) =
-            synchronized(WebPoTokenGenLock) {
-                val shouldRecreate = webPoTokenGenerator == null || forceRecreate ||
-                    webPoTokenGenerator!!.isExpired()
+            webPoTokenGenLock.withLock {
+                val gen = webPoTokenGenerator
```

Comment:
With the `GuardedByMutex` class that I thought up in https://github.com/TeamNewPipe/NewPipe/pull/11829/commits/a8da9946d1c9e531328d6a2a05bedc13b2e6df5e it would be possible to put the generator into the lock, making it impossible to access it without locking first. I think that’s a pretty easy win.